utils/downloads.go

```diff
@@ -0,0 +1,91 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"gopkg.in/cheggaaa/pb.v1"
+)
+
+var snapshots = map[string]string{
+	"mainnet": "https://juno-snapshot.s3.us-east-2.amazonaws.com/mainnet/juno_mainnet_v0.5.0_136902.tar",
+	"goerli":  "https://juno-snapshot.s3.us-east-2.amazonaws.com/goerli/juno_goerli_v0.5.0_839969.tar",
+	"goerli2": "https://juno-snapshot.s3.us-east-2.amazonaws.com/goerli2/juno_goerli2_v0.5.0_135973.tar",
+}
+
+const (
+	downloadBufferSize = 8192 // 8KB
+	dirPermissionMod   = 0o770
+)
+
+type Downloader interface {
+	DownloadAndWrite(ctx context.Context, reader io.Reader, writer io.Writer, bar *pb.ProgressBar) error
+}
+
+type RealDownloader struct{}
+
+func (d *RealDownloader) DownloadAndWrite(ctx context.Context, reader io.Reader, writer io.Writer, bar *pb.ProgressBar) error {
+	buffer := make([]byte, downloadBufferSize)
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("canceled download")
```

Comment:
Isn't anything that prevents from completing the download is considered an error, or if it comes from the caller it is not? Besides that, the process may be ended by someone else, maybe the OS.

Reply:
> if it comes from the caller it is not

Exactly. We don't need to tell the caller that they canceled the context--the caller already knows that. It makes the `if err != nil` idiom hard to use, since the caller always has to check if we're just telling it that we "canceled download".

> the process may be ended by someone else, maybe the OS

That wouldn't cancel the context, those would be fatal errors that would manifest some other way.